bin/specextract

```diff
@@ -485,244 +556,95 @@ def set_badpix(evtfile, bpixfile, instrument, verbose):
 
-def sort_files(file_stack, file_type, key):
+def mk_asphist(asol_param, evt_filename_filter, full_outroot,
+               dtffile, instrument, chip_id, verbose, clobber, tmpdir):
+    """
+    Run asphist to create one aspect histogram file per user-input source
+    observation for input into sky2tdet ('weight=yes') or mkarf ('weight=no');
+    the assumption is that it is appropriate to analyze each observation using
+    only one, and not multiple, aspect histogram files.
 
-    file_count = len(file_stack)
-    files_orig_order = list(range(file_count))
+    Return per chip aspect histogram and file name string
+    """
 
-    for i in range(0, file_count):
-        files_orig_order[i] = file_stack[i]
+    with new_pfiles_environment(ardlib=True,copyuser=False):
+        asp_out = tempfile.NamedTemporaryFile(suffix=f"_asphist{chip_id}",dir=tmpdir)
 
+        asphist.punlearn()
 
-    keyvals_orig_order = get_keyvals(file_stack, key)
+        asphist.infile = asol_param # single file, @files.lis, or
+                                    # comma-separated list of files
 
-    if None in keyvals_orig_order:
-        raise IOError(f"One or more of the entered {file_type} files is missing the required {key} header keyword value. Exiting.")
+        if instrument == "ACIS":
+            asphist.evtfile = f"{evt_filename_filter}[ccd_id={chip_id}]"
+            asphist.dtffile = ""
+        else:
+            asphist.evtfile = f"{evt_filename_filter}[chip_id={chip_id}]"
+            asphist.dtffile = dtffile
 
+        asphist.outfile = asp_out.name
+        asphist.verbose = verbose
+        asphist.clobber = "yes"
 
-    keyvals_sorted = sorted(keyvals_orig_order) #be sure numbers are not strings
+        asphist()
+        
+    return asp_out, asp_out.name
 
-    files_sorted = list(range(file_count))
 
-    for i in range(0, file_count):
-        sort_ind = keyvals_orig_order.index(keyvals_sorted[i])
-        files_sorted[i] = files_orig_order[sort_ind]
 
-    return files_sorted
+def pi_alternate_wmap(pi,wmap):
+    """
+    replace the WMAP in the PI file (typically the dmextract generated version) 
+    with an alternate WMAP (typically generated by sky2tdet or clipped PI WMAP).
+    """
 
+    tmp_spec = tempfile.NamedTemporaryFile(suffix="PI-WMAP",dir=tmpdir)    
+    tmp_wmap = tempfile.NamedTemporaryFile(suffix="wmap",dir=tmpdir)
+    
+    cr = pcr.read_file(pi,mode="rw")
 
-##########################################################################
-# Usage:
-#  group_by_obsid(file_stack, file_type)
-#
-#
-# Aim:
-#  Read the OBS_ID value from the header of each file in the input stack
-#  in order to create a dictionary matching each ObsID to one or a list of
-#  files. Then, this dictionary can be used by other functions in the script
-#  to assign the appropriate file(s) (e.g., asol) to each source observation,
-#  by obsid.
-#
-##########################################################################
+    ds = cr.get_dataset()
```

Comment:
rather than g ofrom crate to dataset you should be able to do sokmething like (untested)

```
ds = pcr.CrateDataset(pi, mode='r')   # I'm guessing about mode=r, it may need to be rw
ds.delete_crate('WMAP')
ds.write(tmp_spec_name, clobber=True)
```

Reply:
This works and I've updated things, albeit this might wind up being unused code that can be dropped, now that I've done some further time testing with WMAP clipping that this functionality may go along with.